src/dos/cdrom_ioctl_linux.cpp

```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
 	return (ret > 0);
 }
 
-bool CDROM_Interface_Ioctl::ReadSectors(PhysPt buffer,
-                                        const bool raw,
-                                        const uint32_t sector,
-                                        const uint16_t num)
+bool CDROM_Interface_Ioctl::GetAudioTracks(uint8_t& stTrack, uint8_t& end, TMSF& leadOut)
+{
+	if (cdrom_fd == -1) {
+#ifdef DEBUG_IOCTL
+		LOG_WARNING("CDROM_IOCTL: GetAudioTracks: cdrom_fd not open");
+#endif
+		return false;
+	}
+	cdrom_tochdr toc = {};
+	if (ioctl(cdrom_fd, CDROMREADTOCHDR, &toc) != 0) {
+#ifdef DEBUG_IOCTL
+		LOG_WARNING("CDROM_IOCTL: GetAudioTracks: ioctl CDROMREADTOCHDR failed");
+#endif
+		return false;
+	}
+	cdrom_tocentry entry = {};
+	entry.cdte_track     = CDROM_LEADOUT;
+	entry.cdte_format    = CDROM_MSF;
+	if (ioctl(cdrom_fd, CDROMREADTOCENTRY, &entry) != 0) {
+#ifdef DEBUG_IOCTL
+		LOG_WARNING("CDROM_IOCTL: GetAudioTracks: ioctl CDROMREADTOCENTRY failed");
+#endif
+		return false;
+	}
+	stTrack     = toc.cdth_trk0;
+	end         = toc.cdth_trk1;
+	leadOut.min = entry.cdte_addr.msf.minute;
+	leadOut.sec = entry.cdte_addr.msf.second;
+	leadOut.fr  = entry.cdte_addr.msf.frame;
+#ifdef DEBUG_IOCTL
+	LOG_INFO("CDROM_IOCTL: GetAudioTracks => start track is %2d, last playable track is %2d, "
+	         "and lead-out MSF is %02d:%02d:%02d",
+	         stTrack,
+	         end,
+	         leadOut.min,
+	         leadOut.sec,
+	         leadOut.fr);
+#endif
+	return true;
+}
+
+bool CDROM_Interface_Ioctl::GetAudioTrackInfo(uint8_t track, TMSF& start,
+                                              unsigned char& attr)
+{
+	if (cdrom_fd == -1) {
```

Comment:
It would be better to have an `is_opened` flag instead of deriving that information from magic values in multiple places. Then this becomes `if (is_opened)`.

Reply:
Could do that I guess.  I think it just makes things more complex.  When you do an `open`, it's always -1 if error.  Anything else is valid.  0 is even valid (it's stdin).

I'd rather do a macro or something if readability is a concern.  Those checks can maybe even be omitted entirely since the `ioctl` will fail and get handled later anyway (though it does avoid potentially expensive syscalls if we know they're going to fail ahead of time).

Reply:
> I wrote an `IsOpen()` function that checks whether the file descriptor is in error state. I'm a little afraid if I make a separate boolean that there could be bugs if one of the variables gets changed but not the other.

It's fine, it's not that complex. We do similar things in a number of places.

> Hopefully this improves readability while keeping the data in one place.

Yes, I think it does. Reduced duplication and gets rid of "deriving" the state from magic values in file descriptors, which although not a huge deal, it's not ideal either.